node/src/components/diagnostics_port/tasks.rs

```diff
@@ -0,0 +1,630 @@
+    CreateTempDir(#[source] io::Error),
+    /// Error trying to create a file in the temporary directory.
+    #[error("could not create file in temporary directory")]
+    CreateTempFile(#[source] io::Error),
+    /// Error trying to reopen the file in the temporary directory after writing.
+    #[error("could not reopen file in temporary directory")]
+    ReopenTempFile(#[source] io::Error),
+}
+
+impl Session {
+    /// Creates a serializer for an `EraDump`.
+    fn create_era_dump_serializer(&self) -> fn(&EraDump<'_>) -> Result<Vec<u8>, Cow<'static, str>> {
+        // TODO: This function could probably be a generic serialization function for any `T`, but
+        // the conversion is tricky due to the lifetime arguments on `EraDump` and has not been done
+        // yet.
+        match self.output {
+            OutputFormat::Interactive => |data: &EraDump| {
+                let mut buf = data.to_string().into_bytes();
+                buf.push(b'\n');
+                Ok(buf)
+            },
+            OutputFormat::Json => |data: &EraDump| {
+                let mut buf = serde_json::to_vec(&data).map_err(|err| {
+                    Cow::Owned(format!("failed to serialize era dump as JSON: {}", err))
+                })?;
+                buf.push(b'\n');
+                Ok(buf)
+            },
+            OutputFormat::Bincode => |data: &EraDump| {
+                bincode::serialize(&data).map_err(|err| {
+                    Cow::Owned(format!("failed to serialize era dump as bincode: {}", err))
+                })
+            },
+        }
+    }
+
+    /// Creates a generic serializer that is writing to a temporary file.
+    ///
+    /// The resulting serializer will write to the given file.
+    fn create_queue_dump_format(&self, file: File) -> QueueDumpFormat {
+        match self.output {
+            OutputFormat::Interactive => QueueDumpFormat::debug(file),
+            OutputFormat::Json => {
+                QueueDumpFormat::serde(FileSerializer::Json(serde_json::Serializer::new(file)))
+            }
+            OutputFormat::Bincode => {
+                QueueDumpFormat::serde(FileSerializer::Bincode(bincode::Serializer::new(
+                    file,
+                    // TODO: Do not use `bincode::serialize` above, but rather always instantiate
+                    // options across the file to ensure it is always the same.
+                    DefaultOptions::new()
+                        .with_fixint_encoding()
+                        .allow_trailing_bytes(),
+                )))
+            }
+        }
+    }
+
+    /// Processes a single command line sent from a client.
+    async fn process_line<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        writer: &mut OwnedWriteHalf,
+        line: &str,
+    ) -> io::Result<bool>
+    where
+        REv: From<DumpConsensusStateRequest> + From<ControlAnnouncement> + Send,
+    {
+        debug!(%line, "line received");
```

Comment:
I know I already reviewed this code. Pity that I have to do it again. Instead, you could have made the changes first and then maybe do a rename as a separate PR.

Reply:
I'm sorry :(